sui-execution/latest/sui-adapter/src/temporary_store.rs

```diff
@@ -186,14 +187,33 @@ impl<'backing> TemporaryStore<'backing> {
         mut self,
         shared_object_refs: Vec<ObjectRef>,
         transaction_digest: &TransactionDigest,
-        transaction_dependencies: Vec<TransactionDigest>,
+        mut transaction_dependencies: BTreeSet<TransactionDigest>,
         gas_cost_summary: GasCostSummary,
         status: ExecutionStatus,
         gas_charger: &mut GasCharger,
         epoch: EpochId,
     ) -> (InnerTemporaryStore, TransactionEffects) {
         let mut modified_at_versions = vec![];
 
+        // Regardless of execution status (including aborts), we insert the previous transaction
+        // for any successfully received objects during the transaction.
+        for (id, expected_version, expected_digest) in &self.receiving_objects {
+            // If the receiving object is in the loaded runtime objects, then that means that it
+            // was actually successfully loaded (so existed, and there was authenticated mutable
+            // access to it). So we insert the previous transaction as a dependency.
+            if let Some(obj_meta) = self.loaded_runtime_objects.get(id) {
+                // Check that the expected version, digest, and owner match the loaded version,
+                // digest, and owner. If they don't then don't register a dependency.
+                // This is because this could be "spoofed" by loading a dynamic object field.
```

Comment:
can you explain this spoofing? shouldn't an attempt to receive a child object fail because the object will be object owned?

Reply:
Basically, due to the way dynamic object fields work, you can have the following sequence of certificates and executions:
* Object `O` at version `n` with object id `oid` is sent to a shared object `S`
* Tx1 creates a cert to receive `O` at version `n` but never actually tries to receive `O`. BUT Tx1 loads a dynamic object field `F` that will reference this object in the future. (Tx1 could not be executed successfully right now). 
* Tx2 creates a cert, and executes a transaction moving `O` to the dynamic object field `F`. `O` now has an `ObjectOwner()` ownership, but importantly the same object id `oid`.
* Tx1 executes. This will load `O` _via the dynamic object field_ at a different version `n + k` (`k > 0`). Because of this `oid` will be in the `loaded_runtime_objects` but at a different version. 
* If we don't have this check that the object in the loaded runtime object metadata was actually received at the specified version (and I added the other checks just to be doubly sure) then we could register a transaction dependency that actually didn't exist. Not a massive deal, but also not quite correct either. 

A test that demonstrates this is here: https://github.com/MystenLabs/sui/blob/f225f5819ced871e968d6c0538698ab0b69b827e/crates/sui-core/src/unit_tests/transfer_to_object_tests.rs#L1531